### DISCUSSION TITLE
Size Limitation for Server Actions in Next.js 14

### DISCUSSION THREAD

**Discussion by cornelvonwyl** [vercel/next.js#57973] [App Router]
https://nextjs.org/docs/app/api-reference/functions/server-actions#size-limitation

How can I increase the size limit for Server Actions in Next.js 14? Since Server Actions are stable, the documentation is outdated.

**Comment by amplicity** [comment]
ditto -- where is this documentation?

**Reply by SidKH** [reply]
https://nextjs.org/docs/app/api-reference/next-config-js/serverActions#bodysizelimit

**Reply by amplicity** [reply]
In next 14, server actions are no longer experimental. Specifying the options stated in those docs does nothing. Output from the server on start reads:

```
 ⚠  Unrecognized key(s) in object: 'serverActions' at "experimental"
 ⚠ See more info here: https://nextjs.org/docs/messages/invalid-next-config
 ⚠ Server Actions are available by default now, `experimental.serverActions` option can be safely removed.
```

I can confirm that specifying the body limit within the options does not work as stated in documentation. Documentation is outdated and broken for next 14.

**Reply by SidKH** [reply]
Enabling server actions via `experimental.serverActions = true` should trigger the warning.
However, changing the `bodySizeLimit` should work fine. 

https://github.com/vercel/next.js/assets/7762253/2ffdb67e-3425-4b98-9b6a-bfd68313be04

**Reply by amplicity** [reply]
I tried your exact settings and get 
<img width="669" alt="Screenshot 2024-02-08 at 1 26 49 PM" src="https://github.com/vercel/next.js/assets/2263838/4061713a-027b-4a34-927c-beb5f358de2c">

However, it does appear to work with some preliminary testing. My next.config is below. Note that I'm on `14.0.1`, vs your `14.0.4`
```
/** @type {import('next').NextConfig} */
const { version } = require("./package.json");
module.exports = {
  env: {
    version,
  },
  reactStrictMode: true,
  experimental: {
    serverActions: {
      bodySizeLimit:"100mb",
    },
  },
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "**",
      },
    ],
  },
};
```

**Reply by abundiko** [reply]
> https://nextjs.org/docs/app/api-reference/next-config-js/serverActions#bodysizelimit

this doesn't work
currently using next 14.2.2

adding the config specified on the docs does not seem to have any effect.
My walkaround was to upload the files on the client side

**Comment by DaveCole** [comment]
Running into a similar issue - it seems that regardless of whether `serverActions: { bodySizeLimit: ... }` is put in `experimental` or in the main config body for 14.1.4 it doesn't seem to matter.  I have a page (using the old pages folder method) that attempts to do a POST via a form with a field that exceeds 1MB in size, and I get this error (the getServerSideProps code doesn't even execute - I tried to throw an error to catch it and it doesn't get reached):

```
web-1       |  ⨯ Error: {"errors":[{"name":"FIELD_SIZE_EXCEEDED","message":"field \"body\" exceeds 0B"}]}
web-1       |     at getProperError (/app/node_modules/next/dist/lib/is-error.js:41:12)
web-1       |     at DevServer.renderToResponseImpl (/app/node_modules/next/dist/server/base-server.js:1904:53)
web-1       |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
web-1       |     at async DevServer.pipeImpl (/app/node_modules/next/dist/server/base-server.js:895:25)
web-1       |     at async NextNodeServer.handleCatchallRenderRequest (/app/node_modules/next/dist/server/next-server.js:269:17)
web-1       |     at async DevServer.handleRequestImpl (/app/node_modules/next/dist/server/base-server.js:791:17)
web-1       |     at async /app/node_modules/next/dist/server/dev/next-dev-server.js:331:20
web-1       |     at async Span.traceAsyncFn (/app/node_modules/next/dist/trace/trace.js:151:20)
web-1       |     at async DevServer.handleRequest (/app/node_modules/next/dist/server/dev/next-dev-server.js:328:24)
web-1       |     at async invokeRender (/app/node_modules/next/dist/server/lib/router-server.js:174:21)
web-1       |     at async handleRequest (/app/node_modules/next/dist/server/lib/router-server.js:353:24)
web-1       |     at async requestHandlerImpl (/app/node_modules/next/dist/server/lib/router-server.js:377:13)
web-1       |     at async Server.requestListener (/app/node_modules/next/dist/server/lib/start-server.js:140:13) {
web-1       |   page: '/admin/customers/1/dashboards/1/edit'
web-1       | }
```

**Comment by amplicity** [comment]
At this point, I'm convinced that there is not an official way to increase serverActions bodySizeLimit. After testing, I don't believe the limit I've put in place above actually works as expected. I followed the instructions [here](https://nextjs.org/docs/app/api-reference/next-config-js/serverActions) for 14, which yields an 'incorrect config' error upon application start, and does not seem to work as expected.

Can someone from the nextjs team comment as to whether this is definitively supported or not? If so, we'd really appreciate the documentation being updated to reflect the best practice.

**Comment by zdaar** [comment]
I have the same issue. Doc has not been updated, and there seem to be no way to do this.

**Comment by DaveCole** [comment]
The workaround I came up with in the meantime is to compress the body of the POST before sending it and decompress it on the API side afterwards.  For strings, I use `lz-string` but not sure what's out there for binaries.

**Comment by vubui** [comment]
How do we flag this to the Next.js team?

I followed the documentation; however, the config seems to have no effect on the body size limit.

**Comment by angelpadillar** [comment]
This is still an ongoing issue in Next.js 14 had to compress and then decompress in the API route to work around.

**Comment by LuisGilGB** [comment]
I have a small PoC in `v15.0.0-canary.68` to test some transcription issues with Whisper and **it looks like the limit of server functions is being applied to server actions when you have the app deployed on Vercel**.

Basically my PoC consists in a local video uploader, the file is then sent on a button click to Whisper to get the transcription. Everything happens in the client except the request to OpenAI, which is made from the server (to keep access keys private). First I implemented it defining an API, and when I hit the 4.5MB size limit of the payload of server functions (I'm dealing with videos that usually are up to 10MB), I used a server action instead. Server actions did fine in local environment; but when deployed the app on Vercel I got the `FUNCTION_PAYLOAD_TOO_LARGE` error again.

**Summing it up, these were my findings:**

Using a Server action to send a file in a form data payload. Size limit of 25MB in Next Config:

| | Local | Vercel Preview |
|-|-------|------|
| 4MB video | Ok | Ok |
| 5MB video | Ok | KO |

**Comment by vaxevanis** [comment]
In next v.15.2.2 this is still not fixed. At the end,  I had to patch files under node_modules/next/dist/compiled/next-server/ to make this work, which is not pretty to do.

**Comment by rmrx7** [comment]
Still not fixed as of 15.3.1
Files above 1mb get ignored in prod

```
  experimental: {
    serverActions: {
      bodySizeLimit: "5mb",
    },
  },
```

**Comment by RichardPickman** [comment]
No fix yet?

**Comment by stealthmode12** [comment]
any update?

**Comment by benjipeto** [comment]
Looks like it's not a priority to make this work, what I don't get, why adding it to the documentation if only works in local? Update the docs or fix the issue, pls... There are dozens of people complaining about this for years in multiple threads..
